sw/device/lib/dif/dif_gpio.c

```diff
@@ -0,0 +1,256 @@
+  return read_register(gpio, offset) & index_to_mask(index);
+}
+#ifdef GTEST_TEST
+bool (*read_register_bit_fptr)(dif_gpio_t const *const gpio,
+                               uint32_t const offset,
+                               uint32_t const index) = read_register_bit;
+#define read_register_bit read_register_bit_fptr
+#endif
+
+static inline void write_register(dif_gpio_t const *const gpio,
+                                  uint32_t const offset, uint32_t const val) {
+  abort_if_invalid(gpio);
+  reg32_write(gpio->base_addr, offset, val);
+}
+#ifdef GTEST_TEST
+void (*write_register_fptr)(dif_gpio_t const *const gpio, uint32_t const offset,
+                            uint32_t const val) = write_register;
+#define write_register write_register_fptr
+#endif
+
+static inline void write_register_masked_access(dif_gpio_t const *const gpio,
+                                                uint32_t const lowerOffset,
+                                                uint32_t const upperOffset,
+                                                uint32_t const mask,
+                                                uint32_t const val) {
+  // Check here once and use |reg32_write()| instead of |write_register()|,
+  // which may result in two calls to |abort_if_invalid()|.
+  abort_if_invalid(gpio);
+  uint32_t const lower_mask = mask & 0x0000FFFFU;
+  uint32_t const upper_mask = mask & 0xFFFF0000U;
+  if (lower_mask) {
+    reg32_write(gpio->base_addr, lowerOffset,
+                (lower_mask << 16) | (val & 0x0000FFFFU));
+  }
+  if (upper_mask) {
+    reg32_write(gpio->base_addr, upperOffset,
+                upper_mask | ((val & 0xFFFF0000U) >> 16));
+  }
+}
+#ifdef GTEST_TEST
+void (*write_register_masked_access_fptr)(
+    dif_gpio_t const *const gpio, uint32_t const lowerOffset,
+    uint32_t const upperOffset, uint32_t const mask,
+    uint32_t const val) = write_register_masked_access;
+#define write_register_masked_access write_register_masked_access_fptr
+#endif
+
+static inline void write_register_bit_masked_access(
+    dif_gpio_t const *const gpio, uint32_t const lowerOffset,
+    uint32_t const upperOffset, uint32_t const index, bool const val) {
+  ptrdiff_t const offset = (index < 16) ? lowerOffset : upperOffset;
+  uint32_t const mask = index_to_mask(index % 16);
+  write_register(gpio, offset, (mask << 16) | (val ? mask : 0U));
+}
+#ifdef GTEST_TEST
+void (*write_register_bit_masked_access_fptr)(
+    dif_gpio_t const *const gpio, uint32_t const lowerOffset,
+    uint32_t const upperOffset, uint32_t const index,
+    bool const val) = write_register_bit_masked_access;
+#define write_register_bit_masked_access write_register_bit_masked_access_fptr
+#endif
+
+void dif_gpio_init(dif_gpio_config_t const *const config, dif_gpio_t *gpio) {
+  // Make sure that input is valid.
+  if (!config || !config->base_addr.inner_ptr || !gpio) {
+    abort();
+  }
+
+  // Configure GPIO device at the given |base_addr|.
```

Comment:
Good point. The programmers guide also has an example that enables rising and falling edge interrupts for a pin, but not a level and edge interrupt at the same time. Since it is not explicitly mentioned, I assumed it would be possible but I will double-check.